slides/advanced-02-model-tuning.qmd

```diff
@@ -0,0 +1,1064 @@
+
+
+. . .
+
+We won't be discussing iterative search methods today. But you can learn more in [AML4TD](https://aml4td.org/chapters/iterative-search.html).
+
+## Grid search
+
+A small grid of points trying to minimize the error via learning rate: 
+
+
+![](images/small_init.svg){fig-align="center" width=60%}
+
+
+## Grid search
+
+In reality, we would probably sample the space more densely: 
+
+![](images/grid_points.svg){fig-align="center" width=60%}
+
+
+## Iterative Search
+
+We could start with a few points and search the space:
+
+![](animations/anime_seq.gif){fig-align="center" width=60%}
+
+# Grid Search {background-color="#FAE093FF"}
+
+## Parameters
+
+-   The tidymodels framework provides pre-defined information on tuning parameters (such as their type, range, transformations, etc.).
+
+-   The `extract_parameter_set_dials()` function extracts these tuning parameters and the info.
+
+<br>
+
+```{r}
+#| label: nnet-param
+nnet_param <- 
+  nnet_wflow |> 
+  extract_parameter_set_dials() 
+nnet_param  
+```  
+
+## Different types of grids `r hexes(c("dials"))`  {.annotation}
+
+
+```{r} 
+#| label: grid-types
+#| echo: false
+#| fig-width: 10
+#| fig-height: 2.7
+#| fig-align: 'center'
+#| out-width: 100%
+
+nnet_plot_param <- parameters(class_weights(), learn_rate())
+
+reg_1 <- grid_regular(nnet_plot_param, levels = c(4, 4)) |> 
+  mutate(type = "Regular (balanced)")
+reg_2 <- grid_regular(nnet_plot_param, levels = c(3, 5)) |> 
+  mutate(type = "Regular (unbalanced)")
+
+irreg_1 <- grid_space_filling(nnet_plot_param, size = 16, type = "uniform") |> 
+  mutate(type = "SFD: Uniform")
+
+set.seed(431)
+irreg_2 <- grid_random(nnet_plot_param, size = 16) |> 
+  mutate(type = "Random")
+irreg_3 <- grid_space_filling(nnet_plot_param, size = 16, type = "latin_hypercube") |> 
+  mutate(type = "SFD: Latin Hypercube")
+
+lvls <- c("Regular (balanced)", "Regular (unbalanced)", "Random", 
+          "SFD: Latin Hypercube", "SFD: Uniform")
+
+grids <- 
+  bind_rows(reg_1, reg_2, irreg_1, irreg_2, irreg_3) |> 
+  mutate(type = factor(type, levels = lvls))
+
+grids |> 
+  ggplot(aes(class_weights, learn_rate)) + 
+  geom_point() + 
+  scale_y_log10() +
+  facet_wrap(~ type, nrow = 1) +
+  labs(x = class_weights()$label, y = learn_rate()$label)
+```
+
+
+[Space-filling designs](https://aml4td.org/chapters/grid-search.html#sec-irregular-grid) (SFD) attempt to cover the parameter space without redundant candidates. We recommend these the most, and they are the default. 
+
+## Extract and update parameters `r hexes(c("dials", "workflows"))`
+
+```{r}
+#| label: update-nnet-param
+nnet_param <- 
+  nnet_param |> 
+  update(
+  class_weights = class_weights(c(1, 50))
+  )
+  
+  nnet_param  
+```  
+
+
+## Grid Search `r hexes(c("dials", "workflows", "tune"))` 
+
+```{r}
+#| label: start-parallel
+#| include: false
+library(mirai)
+daemons(0) # reset to sequential if re-running
```

Comment:
I will petition Charlie to allow us to reset the cores without having to use `daemons(0)`. I added this here since `load_all()` resets back to sequential.